ambry-protocol/src/main/java/com/github/ambry/protocol/AmbryRequests.java

```diff
@@ -683,7 +716,162 @@ public void handleReplicaMetadataRequest(NetworkRequest request) throws IOExcept
+      // referenced as java objects without any need for deserialization.
+      replicateBlobRequest = (ReplicateBlobRequest) ((LocalChannelRequest) request).getRequestInfo().getRequest();
+    } else {
+      replicateBlobRequest = ReplicateBlobRequest.readFrom(new DataInputStream(request.getInputStream()), clusterMap);
+    }
+    long requestQueueTime = SystemTime.getInstance().milliseconds() - request.getStartTimeInMs();
+    long totalTimeSpent = requestQueueTime;
+    long startTime = SystemTime.getInstance().milliseconds();
+    metrics.replicateBlobRequestQueueTimeInMs.update(requestQueueTime);
+    metrics.replicateBlobRequestRate.mark();
+    ReplicateBlobResponse response = null;
+
+    // Get the two parameters from the replicateBlobRequest: the blobId and the remoteHostName
+    BlobId blobId = replicateBlobRequest.getBlobId();
+    String remoteHostName = replicateBlobRequest.getSourceHostName();
+    int remoteHostPort = replicateBlobRequest.getSourceHostPort();
+    DataNodeId remoteDataNode = clusterMap.getDataNodeId(remoteHostName, remoteHostPort);
+    GetResponse getResponse = null;
+    try {
+      if (remoteDataNode == null) {
+        logger.error("ReplicateBlobRequest {} couldn't find the remote host {} {} in the clustermap.", blobId,
+            remoteHostName, remoteHostPort);
+        throw new IOException(
+            "ReplicateBlobRequest " + blobId + " couldn't find the remote host " + remoteHostName + " port "
+                + remoteHostPort);
+      }
+      // ON_DEMAND_REPLICATION_TODO: Add configuration as replicationConfig.replicationConnectionPoolCheckoutTimeoutMs
+      final int connectionPoolCheckoutTimeoutMs = 1000;
+
+      List<PartitionRequestInfo> partitionRequestInfoList = new ArrayList<PartitionRequestInfo>();
+      List<BlobId> blobIds = new ArrayList<BlobId>();
+      blobIds.add(blobId);
+      PartitionRequestInfo partitionInfo = new PartitionRequestInfo(blobId.getPartition(), blobIds);
+      partitionRequestInfoList.add(partitionInfo);
+      // Get the Blob including expired and deleted.
+      final String clientId = ON_DEMAND_REPLICATION_CLIENTID_PREFIX + currentNode.getHostname();
+      GetRequest getRequest = new GetRequest(replicateBlobRequest.getCorrelationId(), clientId, MessageFormatFlags.All,
+          partitionRequestInfoList, GetOption.Include_All);
+
+      // step 1: get the blob from the remote replica
+      ConnectedChannel connectedChannel =
+          connectionPool.checkOutConnection(remoteHostName, remoteDataNode.getPortToConnectTo(),
+              connectionPoolCheckoutTimeoutMs);
+      ChannelOutput channelOutput = connectedChannel.sendAndReceive(getRequest);
+      getResponse = GetResponse.readFrom(channelOutput.getInputStream(), clusterMap);
+      if ((getResponse.getError() != ServerErrorCode.No_Error) || (getResponse.getPartitionResponseInfoList() == null)
```

Comment:
nit: we don't need parentheses around each boolean statement

Reply:
i'd recommend substituting this boolean with a small helper like isSomethingX(). Makes it so much more readable.